File: Sprint7/L/C++/main.cpp
// https://contest.yandex.ru/contest/25596/problems/L/
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int n, m;
    std::cin >> n >> m;
    std::vector<int> weights(n);

    for (int i = 0; i < n; i++) {
        std::cin >> weights[i];
    }

    std::vector<int> dp(m + 1, 0);

    for (int i = 0; i < n; i++) {
        for (int j = m; j >= weights[i]; j--) {
            dp[j] = std::max(dp[j], dp[j - weights[i]] + weights[i]);
        }
    }

    std::cout << dp[m] << std::endl;

    return 0;
}

File: Sprint7/L/Rust/src/main.rs
// https://contest.yandex.ru/contest/25596/problems/L/

fn main() {
    let mut lines = std::io::stdin().lines();
    let n_m = lines
        .next()
        .unwrap()
        .unwrap()
        .split_whitespace()
        .map(|x| x.parse::<usize>().unwrap())
        .collect::<Vec<usize>>();
    let (n, m) = (n_m[0], n_m[1]);

    let weights = lines
        .next()
        .unwrap()
        .unwrap()
        .split_whitespace()
        .map(|x| x.parse::<usize>().unwrap())
        .collect::<Vec<usize>>();

    let mut dp = vec![0; m + 1];

    for i in 0..n {
        for j in (weights[i]..=m).rev() {
            dp[j] = std::cmp::max(dp[j], dp[j - weights[i]] + weights[i]);
        }
    }

    println!("{}", dp[m]);
}
